ninja-core/src/main/java/ninja/template/TemplateEngineFreemarker.java

```diff
@@ -316,12 +317,24 @@ public void invoke(Context context, Result result) {
         
         ResponseStreams responseStreams = context.finalizeHeaders(result);
 
-        try (Writer writer = responseStreams.getWriter()) {
+        try {
             
-            freemarkerTemplate.process(map, writer);
-
-        } catch (Exception e) {
+            // Fully buffer the response so in the case of a template error we can 
+            // return the applications 500 error message. Without fully buffering 
+            // we can't guarantee we haven't flushed part of the response to the
+            // client.
+            
+            StringWriter buffer = new StringWriter(64 * 1024);
```

Comment:
Agreed, I was thinking I would get the behavior nailed down then wrap it up with tests but if you prefer I can maintain test coverage as I go.